README.md
# OpenJMU

[![GitHub stars](https://img.shields.io/github/stars/openjmu/OpenJMU)](https://github.com/openjmu/OpenJMU/stargazers)
[![GitHub forks](https://img.shields.io/github/forks/openjmu/OpenJMU)](https://github.com/openjmu/OpenJMU/network)
[![GitHub issues](https://img.shields.io/github/issues/openjmu/OpenJMU)](https://github.com/openjmu/OpenJMU/issues)

![illustration.png](https://i.loli.net/2020/02/15/BIwubHMLk3DRqfj.png)

使用 [Flutter](https://flutter.dev/) 对 [集大通](http://99.jmu.edu.cn/) 进行重写的客户端。

Rebuild from [JiDaTong](http://99.jmu.edu.cn/) client using [Flutter](https://flutter.dev/).

**注意：由于登录体系不对外公开，如果您非本校（集美大学）成员，以任何方式均无法正常登录。**

**Notice: Due to non-public user service, you can't login to this app
if you are not a member of JMU (JiMei University).**

[更新日志](CHANGELOG.md)

## 为什么用Flutter？ Why flutter ?

Flutter是一套跨平台开发框架，能很好地在接口完备的情况下快速构建所需功能，只需负责 UI 逻辑层。

Flutter is a framework which can cross platform,
it can create features when the apis is supported, and we just need to handle with UI layer.

## 为什么要重写？ Why needs to rebuild ?

原本的应用为外包公司制作，由于学校网络建设策略的原因已经停止维护。
应用在各平台上均有大大小小的问题，影响到用户的正常使用。
更新应用能让师生群体均获得更好的应用体验，同时可以扩展新功能。

The older app is developed by other team,
which has been stop maintained due to school's network construction policy.
It occurs various of problem on each platform, makes users hard to use the app.
By rebuild this app, users can get better experience when using this app,
and new feature can be extended.

## 项目无法登录，有什么参考价值？ What kinds of value while the app cannot login?

目前阶段我们无法提供体验账号。如果你对其中的某个实现或组件或依赖使用感兴趣，
可在 QQ 群 FlutterCandies(181398081) 找到我，咨询相关问题。

We cannot provide any account for test due to policy reason at present stage.
If you're interested in some feature/widget/dependencies usage,
you can find me at `FlutterCandies`(181398081) QQ group or file an issue to ask related questions.

## 8多说，上图！ Let's see some screenshots !

| ![](screenshots/1.png) | ![](screenshots/2.png) | ![](screenshots/3.png) |
| ---------------------- | ---------------------- | ---------------------- |
| ![](screenshots/4.png) | ![](screenshots/5.png) | ![](screenshots/6.png) |
| ![](screenshots/7.png) | ![](screenshots/8.png) | ![](screenshots/9.png) |


CHANGELOG.md
# 版本发布记录

## [1.3.6+271] - [2021-12-14]

[修复]
* 修复部分情况下无法加载 HEIC 图片问题。

[优化]
* 升级至 Flutter 2.8。

## [1.3.5+270] - [2021-11-23]

[优化]
* 规避校内短链服务证书问题。

## [1.3.4+269] - [2021-11-19]

[修复]
* 修复相册倒序的问题。

[优化]
* 基于 Android R (12) 编译；
* 禁用校内网络通道；
* 优化网络构建。

## [1.3.3+267] - [2021-10-14]

[修复]
* 修复由于校内网络通道导致的重复登录问题。

[优化]
* 升级至 Flutter 2.5。

## [1.3.2+265] - [2021-09-02]

[修复]
* 修复部分网页链接无法点击的问题。

## [1.3.1+264] - [2021-08-20]

[优化]
* 忽略由学校证书过期导致的图片服务错误。

[修复]
* 修复浏览图片下拉返回时的报错。

## [1.3.0+263] - [2021-07-21]

[新增]
* 接入校内网络通道的 CAS 统一身份认证。

[优化]
* 默认将自动签到；
* 优化校内网络通道登录错误提示；
* 优化首页加载速度；
* 优化动态详情页评论加载速度；
* 优化自定义课程编辑样式；
* 优化个人页呼出时首页的交互；
* 移除登录视频。

## [1.2.3+259] - [2021-06-10]

[修复]
* 修复 iOS 14.5+ 扫码初始化问题；
* 修复扫码预览缩放问题；
* 修复系统版本读取解析问题。

## [1.2.2+256] - [2021-05-11]

[新增]
* 表情面板新增快捷删除按钮。

[修复]
* 紧急处理校内网络通道导致的无法登录的问题；
* 修复 GIF 动图上传被压缩的问题。

## [1.2.1+251] - [2021-04-18]

[优化]
* 增加网络切换时自动检测是否回落校内网络通道；
* 优化校内网络通道检测时机；
* 优化校内网络通道鉴权失效的更新机制；
* 优化图片压缩参数；
* 优化提到某人窗口的文字提示；
* 优化非校内网站的域名替换规则；
* 优化首页轮询接口的请求次数。

[修复]
* 修复 iOS 打开侧边栏出现的回弹问题；
* 修复使用校内网络通道时编辑自定义课程失败的问题；
* 修复正常课程与自定义课程重叠时自定义课程的编辑问题。

## [1.2.0+247] - [2021-04-12]

[新增]
* 静默回落载入校内网络通道，无论是否有缓存或是否在校内，都可以查看最新的课表信息，或访问应用中心的应用。

[优化]
* 优化侧边栏显示模式，可以从桌面任意位置拉出；
* 优化图片上传策略，进一步确保更多图片能进行上传和正常展示；
* 优化个性签名的编辑和显示；
* 优化后台轮询请求；
* 优化校内通道条件下的课程表获取有效性处理；
* 优化周数处理时机，避免被网络请求阻塞影响；
* 优化成绩解析重试条件。

[修复]
* 修复公告弹窗的布局问题；
* 修复 iOS 的图片上传失败问题；
* 修复上传头像后需要重新启动才能生效的问题；
* 修复集市点赞失败问题；
* 修复集市评论图片查看重复的问题；
* 修复自定义课程的增删改查功能；
* 修复潜在的成绩获取失败问题；
* 修复消息页的公告未解析特殊文字的问题；
* 修复搜索时可能出现的空白问题。

## [1.1.0+236] - [2021-03-22]

[新增]
* 广场超过 3 条的同样转发将折叠以减少占用空间，同样应用于通知页内容；
* 重新开放启动页面设置；
* 底栏重新开放文字显示。

[优化]
* 升级 Flutter 至 2.0.2；
* 优化通知页默认页面判断；
* 优化转发及评论间距；
* 优化课程表显示效果；
* 更新部分图标及 JIMOJI；
* 调整表情面板的默认高度；
* 优化搜索指定用户时的交互；
* 优化 WebView 下载文件的提示内容；
* WebView 开启 Hybrid Composition，大幅优化 Android 的 WebView 性能及体验；
* WebView 支持固化 Cookie，重新打开应用不丢失状态；
* 优化异常捕获展示及交互。

[修复]
* 修复部分被屏蔽或删除的动态用户信息不全导致的解析问题；
* 修复部分弹窗条件判断导致的返回按钮丢失的问题；
* 修复点击集市动态外部评论按钮无响应的问题；
* 修复 iOS 底部悬浮部件的空白问题；
* 修复提到用户时头像状态错位问题；
* 修复个人签名在非 UTF-16 字符下错误的问题；
* 修复嵌套特殊文字的解析问题；
* 修复应用中心常用应用编辑存储问题；
* 修复修改字号后未立即生效的问题；
* 修复 WebView 拉起支付宝无法支付的问题；
* 修复 iOS 更新跳转地址问题；
* 调整用户二维码在深色模式下的匹配率。

## [1.0.0+229] - [2021-03-09]

[新增]
* 全新设计，更快、更便利、更简洁；
* 原固定背景的登录界面更改为视频动态背景；
* LOGO 由手写体更新为印刷体；
* 新增 JIMOJI 表情，纯手绘、更多有趣内容；
* 新增常用应用常驻并更新 2.0 版本应用图标，快速直达应用；
* 长按图片支持识别二维码；
* 重新开放广场带图转发、集市多图评论。

[优化]
* 所有组件性能优化；
* 所有设计细节优化；
* 刷新指示器的悬浮样式更改为融合样式；
* 频繁访问的“广场”与“集市”进行分区，方便操作与消息分类；
* 改进消息的显示方式；
* 优化动态发布页的图片的显示、排序、删除等功能；
* WebView 按键布局逻辑更新，加入切换桌面/手机视图功能。

[修复]
* 修复各类构建的使用、交互、解析问题。

## [0.9.4+176] - [2020-02-22]

[新增]
* 集市动态详情页删除动态时将移除列表对应动态；
* 内置网页浏览重新添加进度条。

[优化]
* 进一步控制搜索内容过滤条件；
* 诚毅学院用户点击“教学评测”将直接跳转至教务系统；
* 搜索页细节优化；
* 扫描二维码为网址时，调整页面栈为推出扫描页后再跳转浏览。

[修复]
* 修复发布动态后不会自动返回首页的问题；
* 修复在集市动态详情删除动态后不会自动返回列表的问题；
* 修复集市详情回复后列表消失的问题；
* 修复网络状态遮罩偶发不显示的问题；
* 修复网络状态遮罩在特定操作下无法去除的问题；
* 修复在启动页下进行特定操作后网络状态提示不正常导致无法尝试登陆的问题；
* 修复动态的原动态类型解析错误问题；
* 修复扫描二维码结果为空时卡死的问题；
* 修复在夜间模式下生成的新版用户二维码无法被识别的问题。

## [0.9.3+172] - [2020-02-20]

[新增]
* 新增“校内资源访问”应用，无忧访问校内应用；
* 集市动态新增删除评论；
* 课程格子新增周数显示，移除“非本周”文字。

[优化]
* 设备未物理连接网络时将显示遮罩提示；
* 优化登录错误的提示；
* 移除集市动态中添加话题的“话题”文字；
* 调整应用中心，应用图标统一；
* 规范用户黑名单操作行为；
* 调整周数规则，周一为教学周第一天（原为周日）；
* 调试信息样式调整；
* 调整用户关注行为反馈；
* 调整用户二维码样式；
* 进一步限制搜索触发条件；
* 禁用转发带图。

[修复]
* 修复注销并重新登录后，再次打开应用仍然无法恢复登录态的问题；
* 修复动态列表未正确触发错误提示的问题；
* 修复动态评论的操作项缺失混乱问题；
* 修复课表中未跟随已选周数变更的日期指示器问题；
* 修复课表对有第11节时的显示位置问题；
* 修复应用为手动夜间模式时部分课程详情显示为灰色的问题；
* 修复课表解析出错时未显示错误提示的问题；
* 修复周数缓存为空时可能造成的写入问题；
* 修复 cookie 无法在内置 webview 中持久化的问题；
* 修复 webview 中部分链接无法跳转的问题；
* 修复用户页查看头像时页面栈推出的问题；
* 修复夜间模式下生成的用户二维码不能被识别的问题；
* 修复用户更改头像后查看大图仍然为旧头像的问题。

## [0.9.2+163] - [2020-02-12]

[新增]
* 增加“问卷调查”图标；
* 首页的公告可以临时关闭；
* 增加研究生、继续教育学生判断，并链接及屏蔽相关功能点。

[优化]
* 安卓移除应用内更新，将直接跳转酷安或浏览器；
* 公告组件细节调整；
* 优化广场和集市的时间显示；
* 调整诚毅学院学生识别规则。

[修复]
* 修复 iOS 部分设备信息获取不完整的问题。

## [0.9.1+161] - [2020-02-06]

[新增]
* 移除启动页延时限制，APP 轻松秒开；
* 课程表、成绩、应用中心数据完全缓存，双倍加速应用体验；
* 夜间模式默认跟随系统，切换无忧；
* 新增隐藏被屏蔽的广场动态设置项，默认开启；
* 全新设计的设置页，功能直达层级清晰；
* 全新设计的关于页，版本履历、官方网站、吐个槽、检查更新，一次看个够；
* 首页底栏更新设计，用户头像将展示在末位；
* 图标焕然一新，风格统一；
* 广场、集市动态操作项使用全新图标并统一样式；
* 图片新增动图标识、优化长图标识；
* 查看图片更清晰，新增下拉返回手势并调整动画，操作更快捷；
* 新增风格统一的加载指示器、确认操作提示框；
* 内建网页浏览调整为传统操作模式，可在更多中复制链接或使用系统浏览器打开，文件可以正常下载；
* 新增清除数据按钮，长按即可抹去一切烦恼；
* 移除对 armv7 的支持。

[优化]
* 重构存储内建，最大程度保持可用率；
* 安卓跳转页面时将使用 Android 10 的过渡动画；
* 获取公告、更新信息、周数失败时将自动重试；
* 优化更新弹窗设计；
* 调整发布动态所需要的系统权限；
* 调整布局以解决顶栏阴影切割问题；
* 状态切换变动的布局添加淡化效果；
* 调整 Tab 页指示器以统一设计；
* 广场动态查看详情时更多操作移至顶栏；
* 广场举报将检查是否重复；
* 优化点赞数值钳制；
* 评论内图片移除预览，显示为“查看图片”；
* 通知的评论增加图片识别；
* 移除对图片网格的大小限制；
* 优化搜索判断及交互逻辑；
* 当前周非学期内时，课程将不再显示为灰色，便于课程截图及分享；
* 长按应用可选择是否使用系统浏览器打开；
* 针对夜间模式优化启动页和登录页；
* 优化个人页面问候语内容；
* 优化加号页、通知入口页显示设计及交互钳制；
* 进入通知提醒页将优先切换至有通知的页面；
* 优化修改签名流程；
* 内建课程表与 Web 课程表进一步统一操作行为；
* 重构内建浏览器，支持唤醒已安装的应用；
* 针对分享的微信公众号文章，优化 Open Graph Meta 标签识别；
* 调整部件 getter 与设置内容的同步。

[修复]
* 修复公告部件异常位置的问题；
* 修复扫描二维码选择不包含二维码的图片时，应用无响应的问题；
* 修复 iOS 无法打开广场动态网页链接并闪退的问题；
* 修复 iOS 提示更新时，应用商店跳转至不存在的应用的问题；
* 修复由于应用中心应用筛选导致的诚毅学院用户无法接收到教务通知的问题；
* 修复搜索页头像异常切割的问题；
* 修复格式不正确导致的应用消息日期显示问题。

## [0.8.1+129] - [2019-12-30]

[修复]
* 修复由于错误的API导致的多个课程显示失败；
* 修复空消息阅读及结构化消息的预处理；
* 修复由于特殊文字的构建条件未应用导致的解析问题；
* 修复安卓更新时无法自动安装的问题。

## [0.8.0+128] - [2019-12-23]

[新增]
* 新增应用消息，用户可以收到由应用中心的应用发来的消息；
* 课程表左侧课程节数将同时显示课时；
* 消息页的分区通知移动至首页，点击右上角消息图标即可查看分区通知。

[优化]
* Flutter SDK 升级；
* 编译版本降至28；
* 重构屏幕适配方案，同时调整部件、字体大小；
* 重构部分页面顶栏、分区指示器；
* 重构应用获取模式，以适配应用消息；
* 重构动态内容详情页；
* 重构主题、夜间模式构建方式，将消耗更低的性能；
* 重构周数控制器，以适配课表；
* 启动页30秒未成功登录将强制跳转至登录页；
* 首页底部入口去除文字；
* 调整动态内容的日期显示；
* 调整加号页动画及细节；
* 调整加载更多的样式；
* 课程表添加错误提示；
* 动态内容为4图时将在单行内显示；
* 搜索内容返回唯一用户时将跳转至用户页；
* A屏黑将和实际夜间模式同步可变更状态。

[修复]
* 修复错误的长图识别；
* 修复搜索到的相关动态为空时同时不展示用户的问题；
* 修复自定义课程返回时的黑屏问题。

## [0.7.0+112] - [2019-11-24]

[新增]
* 新增“集市”，适用公约规定的内容发布；
* 新增根据系统夜间模式自动切换夜间模式的设置 (iOS 13 & Android Q)；
* 新增强制更新，当需要强制更新时，将无法继续使用旧版本；
* 成绩页面新增“教学评测”提示；
* 发布动态时提示确认内容是否符合公约；
* 为开发团队新增了标识；
* iOS现在也将检查更新，并提示跳转至App Store更新。

[优化]
* 调整了链接文字的识别，剔除双字节字符识别从而使网址完整显示不吞字；
* 从加号页中移除了扫描二维码；
* 加号页内图标细节调整；
* 调整了动态的图片布局大小；
* 调整了发布动态选择图片后对图片项操作的交互；
* 首页中移除“关注”和“新闻”，将在后续以其他形式提供；
* “首页->首页”更改为“微博广场”，“+->动态”更改为“微博”，“登录页->查询工号”更改为“学工号查询”；
* 登录页《用户协议》提示调整，按钮锁定样式调整；
* 末行统一内容调整。

[修复]

lib/main.dart
import 'dart:async';
import 'dart:io';

import 'package:connectivity_plus/connectivity_plus.dart';
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';
import 'package:flutter/services.dart';
import 'package:flutter_screenutil/flutter_screenutil.dart' hide SizeExtension;

import 'constants/constants.dart' hide PageRouteType;
import 'openjmu_route.dart';
import 'pages/no_network_page.dart';
import 'pages/no_route_page.dart';
import 'pages/splash_page.dart';

void main() {
  runZonedGuarded<void>(
    () async {
      WidgetsFlutterBinding.ensureInitialized();

      // In order to compare the default avatar locally, here I decide to compare
      // the data of the avatar with a local one.
      rootBundle.load(R.ASSETS_AVATAR_PLACEHOLDER_152_PNG).then((ByteData bd) {
        Instances.defaultAvatarData = bd.buffer.asUint8List().toString();
      });

      await SystemChrome.setPreferredOrientations(<DeviceOrientation>[
        DeviceOrientation.portraitUp,
        DeviceOrientation.portraitDown,
      ]);

      SystemChrome.setSystemUIOverlayStyle(SystemUiOverlayStyle.dark.copyWith(
        statusBarColor: Colors.transparent,
      ));

      await Hive.initFlutter();
      await HiveBoxes.openBoxes();
      await Future.wait(
        <Future<void>>[
          DeviceUtils.initDeviceInfo(),
          PackageUtils.initPackageInfo(),
          NetUtils.initConfig(),
        ],
        eagerError: true,
      );
      NotificationUtils.initSettings();
      _customizeErrorWidget();

      runApp(OpenJMUApp());
    },
    (Object e, StackTrace s) => LogUtils.e(
      'Caught unhandled exception: $e',
      stackTrace: s,
    ),
  );
}

class OpenJMUApp extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => OpenJMUAppState();
}

class OpenJMUAppState extends State<OpenJMUApp> with WidgetsBindingObserver {
  StreamSubscription<ConnectivityResult> connectivitySubscription;

  Brightness get _platformBrightness =>
      Screens.mediaQuery.platformBrightness ?? Brightness.light;

  ToastFuture connectivityToastFuture;

  @override
  void initState() {
    super.initState();

    LogUtils.d('Current platform is: ${Platform.operatingSystem}');
    WidgetsBinding.instance.addObserver(this);
    tryRecoverLoginInfo();

    SchedulerBinding.instance.addPostFrameCallback((_) {
      Connectivity().checkConnectivity().then(connectivityHandler);
    });
    connectivitySubscription =
        Connectivity().onConnectivityChanged.listen(connectivityHandler);

    Instances.eventBus
      ..on<TicketGotEvent>().listen((TicketGotEvent event) {
        MessageUtils.initMessageSocket();
        if (currentUser.isTeacher != true) {
          if (currentUser.isPostgraduate != true) {
            currentContext.read<CoursesProvider>().initCourses();
            currentContext.read<ScoresProvider>().initScore();
          }
        }
        currentContext.read<MessagesProvider>().initMessages();
        currentContext.read<NotificationProvider>().initNotification();
        currentContext.read<ReportRecordsProvider>().initRecords();
//        currentContext.read<SettingsProvider>().getCloudSettings();
        currentContext.read<SignProvider>().getSignStatus();
        currentContext.read<WebAppsProvider>().initApps();
        if (UserAPI.backpackItemTypes.isEmpty) {
          UserAPI.getBackpackItemType();
        }
        UserAPI.initializeBlacklist();
        initPushService();
      })
      ..on<LogoutEvent>().listen((LogoutEvent event) {
        navigatorState.pushNamedAndRemoveUntil(
          Routes.openjmuLogin.name,
          (_) => false,
        );
        if (currentUser.isTeacher != true) {
          if (currentUser.isPostgraduate != true) {
            currentContext.read<CoursesProvider>().unloadCourses();
            currentContext.read<ScoresProvider>().unloadScore();
          }
        }
        currentContext.read<MessagesProvider>().unloadMessages();
        currentContext.read<NotificationProvider>().stopNotification();
        currentContext.read<ReportRecordsProvider>().unloadRecords();
        currentContext.read<SignProvider>().resetSignStatus();
        currentContext.read<WebAppsProvider>().unloadApps();
        UserAPI.backpackItemTypes.clear();
        Future<void>.delayed(250.milliseconds, () {
          currentContext.read<ThemesProvider>().resetTheme();
          currentContext.read<SettingsProvider>().reset();
        });
        DataUtils.logout();
      })
      ..on<FontScaleUpdateEvent>().listen((_) {
        if (mounted) {
          _rebuildAllChildren(context);
        }
      })
      ..on<HasUpdateEvent>().listen(PackageUtils.showUpdateDialog);
  }

  @override
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    connectivitySubscription?.cancel();
    super.dispose();
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    LogUtils.d('AppLifecycleState change to: ${state.toString()}');
    if (state == AppLifecycleState.resumed &&
        Instances.appLifeCycleState != AppLifecycleState.resumed) {
      Future<void>.delayed(
        const Duration(seconds: 1),
        DeviceUtils.setHighestRefreshRate,
      );
      currentContext.read<NotificationProvider>().initNotification();
    } else {
      currentContext.read<NotificationProvider>().stopNotification();
    }
    Instances.appLifeCycleState = state;
  }

  @override
  void didChangePlatformBrightness() {
    if (mounted) {
      setState(() {});
    }
  }

  Future<void> tryRecoverLoginInfo() async {
    if (DataUtils.isLogin()) {
      DataUtils.recoverLoginInfo();
    } else {
      Instances.eventBus.fire(TicketFailedEvent());
    }
    if (mounted) {
      setState(() {});
    }
  }

  void connectivityHandler(ConnectivityResult result) {
    NetUtils.testClassKit();
    checkIfNoConnectivity(result);
    Instances.eventBus.fire(ConnectivityChangeEvent(result));
    Instances.connectivityResult = result;
    LogUtils.d('Current connectivity: $result');
  }

  void checkIfNoConnectivity(ConnectivityResult result) {
    if (result == ConnectivityResult.none) {
      connectivityToastFuture ??= showNoNetworkPage(context);
    } else {
      connectivityToastFuture?.dismiss(showAnim: true);
      if (connectivityToastFuture != null) {
        connectivityToastFuture = null;
      }
    }
  }

  void initPushService() {
    // Init this service only on iOS.
    if (!Platform.isIOS) {
      return;
    }
    try {
      final Map<String, dynamic> data = <String, dynamic>{
        'token': DeviceUtils.devicePushToken,
        'date': DateFormat('yyyy/MM/dd HH:mm:ss', 'en').format(DateTime.now()),
        'uid': currentUser.uid,
        'name': currentUser.name ?? currentUser.uid,
        'workid': currentUser.workId ?? currentUser.uid,
        'buildnumber': PackageUtils.buildNumber,
        'uuid': DeviceUtils.deviceUuid,
        'platform': Platform.isIOS ? 'ios' : 'android',
      };
      LogUtils.d('Push data: $data');
      NetUtils.post<void>(API.pushUpload, data: data).then((dynamic _) {
        LogUtils.d('Push service info upload success.');
      }).catchError((dynamic e) {
        LogUtils.e('Push service upload error: $e', withStackTrace: false);
      });
    } catch (e) {
      LogUtils.e('Push service init error: $e');
    }
  }

  ToastFuture showNoNetworkPage(BuildContext context) {
    return showToastWidget(
      const NoNetworkPage(),
      duration: 999.weeks,
      handleTouch: true,
      position: ToastPosition.center,
    );
  }

  @override
  Widget build(BuildContext context) {
    return MultiProvider(
      providers: providers,
      child: Consumer2<ThemesProvider, SettingsProvider>(
        builder: (
          _,
          ThemesProvider themesProvider,
          SettingsProvider settingsProvider,
          __,
        ) {
          final bool isDark = themesProvider.platformBrightness
              ? _platformBrightness == Brightness.dark
              : themesProvider.dark;
          final ThemeData theme =
              (isDark ? themesProvider.darkTheme : themesProvider.lightTheme)
                  .copyWith(
            pageTransitionsTheme: const PageTransitionsTheme(
              builders: <TargetPlatform, PageTransitionsBuilder>{
                TargetPlatform.android: ZoomPageTransitionsBuilder(),
              },
            ),
          );
          return AnnotatedRegion<SystemUiOverlayStyle>(
            value:
                isDark ? SystemUiOverlayStyle.light : SystemUiOverlayStyle.dark,
            child: Theme(
              data: theme,
              child: OKToast(
                position: ToastPosition.bottom,
                textPadding: const EdgeInsets.symmetric(
                  horizontal: 16,
                  vertical: 12,
                ),
                textStyle: const TextStyle(fontSize: 14),
                child: MaterialApp(
                  navigatorKey: Instances.navigatorKey,
                  builder: (BuildContext c, Widget w) {
                    ScreenUtil.init(c, allowFontScaling: true);
                    Widget widget = ScrollConfiguration(
                      behavior: const NoGlowScrollBehavior(),
                      child: NoScaleTextWidget(child: w),
                    );
                    if (Platform.isIOS && Screens.topSafeHeight >= 42) {
                      widget = Stack(
                        children: <Widget>[
                          Positioned.fill(child: widget),
                          const _HiddenLogo(),
                        ],
                      );
                    }
                    return RepaintBoundary(
                      key: Instances.appRepaintBoundaryKey,
                      child: widget,
                    );
                  },
                  title: 'OpenJMU',
                  theme: theme,
                  home: const SplashPage(),
                  navigatorObservers: <NavigatorObserver>[
                    Instances.routeObserver,
                  ],
                  onGenerateRoute: (RouteSettings settings) => onGenerateRoute(
                    settings: settings,
                    getRouteSettings: getRouteSettings,
                    notFoundWidget: NoRoutePage(route: settings.name),
                  ),
                  localizationsDelegates: Constants.localizationsDelegates,
                  supportedLocales: Constants.supportedLocales,
                ),
              ),
            ),
          );
        },
      ),
    );
  }
}

class _HiddenLogo extends StatelessWidget {
  const _HiddenLogo({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return PositionedDirectional(
      start: 0,
      end: 0,
      top: 0,
      height: Screens.topSafeHeight,
      child: Center(
        child: SvgPicture.asset(
          R.IMAGES_OPENJMU_LOGO_TEXT_SVG,
          color: defaultLightColor,
          height: Screens.topSafeHeight * 0.3,
        ),
      ),
    );
  }
}

void _rebuildAllChildren(BuildContext context) {
  void rebuild(Element el) {
    el.markNeedsBuild();
    el.visitChildren(rebuild);
  }

  (context as Element).visitChildren(rebuild);
}

void _customizeErrorWidget() {
  ErrorWidget.builder = (FlutterErrorDetails details) {
    return Material(
      type: MaterialType.transparency,
      child: Container(
        margin: EdgeInsets.all(16.w),
        padding: EdgeInsets.all(30.w),
        decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(13.w),
          color: currentTheme.colorScheme.secondary.withOpacity(0.125),
        ),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            SvgPicture.asset(
              R.ASSETS_PLACEHOLDERS_NO_NETWORK_SVG,
              width: 50.w,
              color: currentTheme.iconTheme.color,
            ),
            VGap(20.w),
            Text(
              '出现了不可预料的错误 (>_<)',
              style: TextStyle(
                color: currentTheme.textTheme.caption.color,
                fontSize: 22.sp,
              ),
            ),
            VGap(10.w),
            Text(
              details.exception.toString(),
              style: TextStyle(
                color: currentTheme.textTheme.caption.color,
                fontSize: 20.sp,
                fontWeight: FontWeight.bold,
              ),
            ),
            VGap(10.w),
            Text(
              details.stack.toString(),
              style: TextStyle(
                color: currentTheme.textTheme.caption.color,
                fontSize: 16.sp,
              ),
              maxLines: 14,
              overflow: TextOverflow.ellipsis,
            ),
            VGap(20.w),
            Tapper(
              onTap: _takeAppScreenshot,
              child: Container(
                padding: EdgeInsets.all(16.w),
                decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(13.w),
                  color: currentTheme.colorScheme.secondary,
                ),
                child: Text(
                  '保存当前位置错误截图',
                  style: TextStyle(
                    color: adaptiveButtonColor(),
                    fontSize: 20.sp,
                    height: 1.24,
                    fontWeight: FontWeight.bold,
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  };
}

Future<void> _takeAppScreenshot() async {
  try {
    final ByteData byteData = await obtainScreenshotData(
      Instances.appRepaintBoundaryKey,
    );
    await PhotoManager.editor.saveImage(byteData.buffer.asUint8List());
    showToast('截图保存成功');
  } catch (e) {
    LogUtils.e('Error when taking app\'s screenshot: $e');
    showCenterErrorToast('截图保存失败');
  }
}
